### PR TITLE
Alesia Yakutsionak - 0

### PR DESCRIPTION
**Pull request by als6k**
## Name
Valentine Zavadskiy

## Homework#
0

## Link to video with demo
https://www.youtube.com/watch?v=ecIWPzGEbFc

## Comment
Level 1,2 and 3.

### CODE REVIEW

File: Alesia Yakutsionak/0/run.rb [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+class ProductPrice
+  attr_accessor :min_price, :max_price, :last_price, :min_price_date, :max_price_date
+
+  def initialize(price, date)
+    @min_price = price
+    @min_price_date = date
+    @max_price = price
+    @max_price_date = date
+    @last_price = price if date == '2018/10'
```
Comment (RGBD) [inline Alesia Yakutsionak/0/run.rb]: make 2018/10 constant, or select greatest date from all available files.

File: Alesia Yakutsionak/0/run.rb [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  end
+
+  def update(price, date)
+    @last_price = price if date == '2018/10'
```
Comment (RGBD) [inline Alesia Yakutsionak/0/run.rb]: make '2018/10' constant

File: Alesia Yakutsionak/0/run.rb [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+      next if i < 8 || row[0].nil?
+      name = clean_name(row[0])
+      price = row[6]
+      add_product(name, price, date)
+    end
+  end
+
+  def collect_from_xlsx(file)
+    xlsx = Roo::Excelx.new(file)
+    date = parse_date(xlsx.sheet(0).row(3).compact.first)
```
Comment (RGBD) [inline Alesia Yakutsionak/0/run.rb]: make 3 constant

File: Alesia Yakutsionak/0/run.rb [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    puts "Can't parse date in #{file}" unless date
+
+    xlsx.sheet(0).each_row_streaming(offset: 7) do |row|
```
Comment (RGBD) [inline Alesia Yakutsionak/0/run.rb]: make 7 constant

File: Alesia Yakutsionak/0/run.rb [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    name = clean_name(row[0].cell_value)
+      next if name.nil?
+      price = row[6].cell_value
+      add_product(name, price, date)
+    end
+  end
+
+  def add_product(name, price, date)
+    price = price.to_f
+    price /= 10_000 if date < '2017/01'
```
Comment (RGBD) [inline Alesia Yakutsionak/0/run.rb]: make 10000, 2017/01 constants
also you can add helper method
```
def convert_price(price, date)
...
price = convert_price(price, date)
```
that will handle both denomination and rounding

File: Alesia Yakutsionak/0/run.rb [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    existing_product = products[name]
+    if existing_product
+      existing_product.update(price, date)
+    else
+      products[name] = ProductPrice.new(price, date)
+    end
+  end
+
+  def parse_date(date_str)
+    month = MONTHS.find{|m| date_str.include?(m)}
```
Comment (RGBD) [inline Alesia Yakutsionak/0/run.rb]: month_number = MONTH.index { ... }

File: Alesia Yakutsionak/0/run.rb [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    return unless month
+    month_number = MONTHS.index(month) + 1
+    match_data = date_str.match(/(\d{4})/)
```
Comment (RGBD) [inline Alesia Yakutsionak/0/run.rb]: ```year = date_str[/\d{4}/]```

File: Alesia Yakutsionak/0/run.rb [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+    product_price = products[product_name]
+    puts "#{product_name} is #{product_price.last_price} BYN in Minsk these days."
+    puts "Lowest was on #{product_price.min_price_date} at price #{product_price.min_price} BYN"
+    puts "Maximum was on #{product_price.max_price_date} at price #{product_price.max_price} BYN"
+  end
+
+  def show_same_price_products(found, product_name)
+    product_price = products[product_name]
+    same_price_products = find_same_price(found, product_price.last_price)
+    unless same_price_products.empty?
```
Comment (RGBD) [inline Alesia Yakutsionak/0/run.rb]: return if same_price_products.empty?

**Comment by als6k** [comment]
У меня осталась только эта ошибка:
"Some files could not be reviewed due to errors:
Error: We found some problems with your configuration file", но она есть практически у всех и проблема не в моем коде, как я понимаю.
Все остальные замечания я исправила.
Не могли бы вы пересмотреть мою работу?

**Comment by RGBD** [comment]
Sorry, I expected @yauhenininjia will take over, since he self-assigned this issue.

File: Alesia Yakutsionak/0/run.rb [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+  DATE_ROW = 3
+  FIRST_DATA_ROW = 8
+  PRODUCT_NAME_COL = 0
+  MINSK_PRICE_COL = 6
+
+  def self.collect_data(&block)
+    print 'Please wait'
+    Dir['data/*'].each do |f|
+      if f.end_with?('.xls')
+        collect_from_xls(f, block)
```
Comment (RGBD) [inline Alesia Yakutsionak/0/run.rb]: There is Roo::Spreadsheet with unified interface, so you don't even need to check file extensions and two different methods for parsing

File: Alesia Yakutsionak/0/run.rb [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+  DENOMINATION_VALUE = 10_000
+
+  attr_accessor :products
+
+  def initialize
+    @products = {}
+  end
+
+  def start
+    PriceCollector.collect_data do |name, price, date|
```
Comment (RGBD) [inline Alesia Yakutsionak/0/run.rb]: You were one step from using Enumerator)
Read [this](https://blog.arkency.com/2014/01/ruby-to-enum-for-enumerator/) if you want to know more.

**Pull request merged**